akka-http-core/src/main/scala/akka/http/impl/engine/http2/Http2Demux.scala

```diff
@@ -357,25 +360,50 @@ private[http2] abstract class Http2Demux(http2Settings: Http2CommonSettings, ini
         }
       })
 
+      // -----------------------------------------------------------------
       // FIXME: What if user handler doesn't pull in new substreams? Should we reject them
-      //        after a while or buffer only a limited amount? We should also be able to
-      //        keep the buffer limited to the number of concurrent streams as negotiated
-      //        with the other side.
+      //        after a while or buffer only a limited amount?
       val bufferedSubStreamOutput = new BufferedOutlet[Http2SubStream](substreamOut)
       override def dispatchSubstream(initialHeaders: ParsedHeadersFrame, data: Source[Any, Any], correlationAttributes: Map[AttributeKey[_], _]): Unit =
         bufferedSubStreamOutput.push(Http2SubStream(initialHeaders, data, correlationAttributes))
 
+      // -----------------------------------------------------------------
+      private var completing = false
```

Comment:
Do we still need the extra flag? Could we use `isClosed(substreamIn)` instead?

Reply:
I've tried it and it doesn't work. IIUC, `isClosed(inlet)` only returns true _after_ the invoking `complete(inlet)`.

I think the flag and `tryComplete` are necessary. `isClosed(substreamIn)` won't make it because `completing` (the flag) is already `true` when `isClosed(substreamIn)` is still false. While `completing` is already `true` we still need to have the inlet open to push elements.

Reply:
That part of the complexity is also due to the handling of the `frameIn` `onUpstreamFinish` handler. Only when called from there it can be that `completing && !isClosed(substreamIn)`.